src/Tools/dotnet-stack/Symbolicate.cs

```diff
@@ -0,0 +1,419 @@
+        }
+
+        private static string GetRegex(string line, List<string> xmlList)
+        {
+            string ret = line;
+            string logtagStr = Regex.Match(line, "(?<splitdata>.*?) at").Groups["splitdata"].Value;
+            if (logtagStr.Length != 0)
+            {
+                ret = line.Replace(logtagStr, "");
+            }
+            string typeMethodStr = Regex.Match(line, "at (?<splitdata>.*?)\\((.*)\\)").Groups["splitdata"].Value;
+            string methodStr = typeMethodStr.Split(".")[typeMethodStr.Split(".").Length - 1];
+            string typenameStr = typeMethodStr.Replace("." + methodStr, "");
+            string parameterStr = Regex.Match(line, methodStr + "\\((?<splitdata>.*?)\\)").Groups["splitdata"].Value;
+            string assemblyStr = Regex.Match(line, " in (?<splitdata>.*?)\\: ").Groups["splitdata"].Value;
+            string[] tokenOffsetStr = Regex.Match(line, "\\: token (?<splitdata>.*)?").Groups["splitdata"].Value.Split("+");
+            string xmlStr = assemblyStr.Contains(".ni.dll") ? assemblyStr.Replace(".ni.dll", ".xml") : assemblyStr.Replace(".dll", ".xml");
+
+            if (tokenOffsetStr.Length != 2 || methodStr == "" || typenameStr == "" || assemblyStr == "")
+            {
+                return ret;
+            }
+
+            StackTraceInfo stInfo = new StackTraceInfo() { Type = typenameStr, Method = methodStr, Assembly = assemblyStr, Token = tokenOffsetStr[0], Offset = tokenOffsetStr[1] };
+
+            foreach (var xmlPath in xmlList)
+            {
+                if (xmlPath.Contains(xmlStr))
+                {
+                    GetLineFromXml(xmlPath, stInfo);
+                    if (stInfo.Filepath != null && stInfo.StartLine != null)
+                    {
+                        ret = $" at {stInfo.Type}.{stInfo.Method}({parameterStr}) in {stInfo.Filepath}:line {stInfo.StartLine}";
+                        break;
+                    }
+                }
+            }
+            return ret;
+        }
+
+        private static void GetLineFromXml(string xmlPath, StackTraceInfo stInfo)
+        {
+            try
+            {
+                XmlDocument xmlDoc = new XmlDocument();
+                xmlDoc.Load(xmlPath);
+                XmlElement xRoot = xmlDoc.DocumentElement;
+                XmlNodeList xnList = xRoot.ChildNodes;
+                int xnCount = xnList.Count;
+                if (xnCount > 0)
+                {
+                    for (int i = xnCount - 1; i >= 0; i--)
+                    {
+                        XmlNode node = xnList[i];
+                        if (node.Name == "files")
+                        {
+                            ParseFile(node.ChildNodes, stInfo);
+                        }
+                        else if (node.Name == "methods")
+                        {
+                            ParseMethod(node.ChildNodes, stInfo);
+                        }
+                    }
+                }
```

Comment:
Can we cache this load and the symbol lookup? This looks like it's going to load the file, parse the XML, and walk every method in the parsed XML for every frame of a stack trace. If a stack has many frames from a single assembly with a large PDB, this could be expensive.